File: Payroll.API/Models/RecruitmentCases.cs
namespace Payroll.API.Models;

public class RecruitmentWorkOrder
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public string ClientName { get; set; } = "";
    public string WorkOrderNumber { get; set; } = "";
    public DateTime ReceivedAtUtc { get; set; }
    public string ReceivedFrom { get; set; } = "";
    public string Subject { get; set; } = "";
    public string Remarks { get; set; } = "";
    public string Status { get; set; } = "Draft";
    public int OverallSlaMinutes { get; set; }
    public DateTime? DueAtUtc { get; set; }
    public int CreatedByUserId { get; set; }
    public DateTime CreatedAtUtc { get; set; }
    public DateTime UpdatedAtUtc { get; set; }
    public int LineCount { get; set; }
    public int OpenCaseCount { get; set; }
    public List<RecruitmentWorkOrderLine> Lines { get; set; } = [];
}

public class RecruitmentWorkOrderLine
{
    public long Id { get; set; }
    public long WorkOrderId { get; set; }
    public int LineNumber { get; set; }
    public string PositionName { get; set; } = "";
    public string PayBandLevelCode { get; set; } = "";
    public int NumberOfPositions { get; set; } = 1;
    public string Location { get; set; } = "";
    public string Division { get; set; } = "";
    public long? RequisitionId { get; set; }
    public long? PositionId { get; set; }
    public string Status { get; set; } = "Open";
    public DateTime CreatedAtUtc { get; set; }
    public DateTime UpdatedAtUtc { get; set; }
}

public class SaveRecruitmentWorkOrder
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public string WorkOrderNumber { get; set; } = "";
    public DateTime ReceivedAtUtc { get; set; }
    public string ReceivedFrom { get; set; } = "";
    public string Subject { get; set; } = "";
    public string Remarks { get; set; } = "";
    public string Status { get; set; } = "Draft";
    public int OverallSlaMinutes { get; set; }
    public List<SaveRecruitmentWorkOrderLine> Lines { get; set; } = [];
}

public class SaveRecruitmentWorkOrderLine
{
    public long Id { get; set; }
    public int LineNumber { get; set; }
    public string PositionName { get; set; } = "";
    public string PayBandLevelCode { get; set; } = "";
    public int NumberOfPositions { get; set; } = 1;
    public string Location { get; set; } = "";
    public string Division { get; set; } = "";
    public long? RequisitionId { get; set; }
    public long? PositionId { get; set; }
    public string Status { get; set; } = "Open";
}

public class RecruitmentHiringCase
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public string ClientName { get; set; } = "";
    public long WorkOrderId { get; set; }
    public string WorkOrderNumber { get; set; } = "";
    public long WorkOrderLineId { get; set; }
    public long? RequisitionId { get; set; }
    public long? PositionId { get; set; }
    public string PositionName { get; set; } = "";
    public string PayBandLevelCode { get; set; } = "";
    public string Division { get; set; } = "";
    public long PipelineVersionId { get; set; }
    public string PipelineName { get; set; } = "";
    public DateTime SlaAnchorAtUtc { get; set; }
    public DateTime? OverallDueAtUtc { get; set; }
    public long? CurrentStageInstanceId { get; set; }
    public string CurrentStageName { get; set; } = "";
    public string CurrentStakeholderCode { get; set; } = "";
    public string Status { get; set; } = "Active";
    public string AdvanceStatus { get; set; } = "";
    public long? AdvanceRequestId { get; set; }
    public string AdvanceMessage { get; set; } = "";
    public DateTime StartedAtUtc { get; set; }
    public DateTime UpdatedAtUtc { get; set; }
    public List<RecruitmentHiringCaseStage> Stages { get; set; } = [];
}

public class StartRecruitmentHiringCaseRequest
{
    public long WorkOrderLineId { get; set; }
    public long PipelineVersionId { get; set; }
}

public class RecruitmentHiringCaseStage
{
    public long Id { get; set; }
    public long HiringCaseId { get; set; }
    public long PipelineStageId { get; set; }
    public string StageCode { get; set; } = "";
    public string StageName { get; set; } = "";
    public int DisplayOrder { get; set; }
    public string StakeholderCode { get; set; } = "";
    public int? TargetOffsetMinutes { get; set; }
    public bool AllowPause { get; set; } = true;
    public string PauseBehavior { get; set; } = "ShiftStageAndOverall";
    public bool RequiresApproval { get; set; }
    public long? ApprovalWorkflowId { get; set; }
    public bool IsTerminal { get; set; }
    public string Status { get; set; } = "Pending";
    public DateTime? EnteredAtUtc { get; set; }
    public DateTime? DueAtUtc { get; set; }
    public DateTime? CompletedAtUtc { get; set; }
    public long PausedDurationSeconds { get; set; }
    public bool IsPaused { get; set; }
    public bool IsSlaBreached { get; set; }
    public List<RecruitmentHiringCasePausePeriod> PauseHistory { get; set; } = [];
    public List<RecruitmentStageProcessDocumentRequirement> ProcessDocumentRequirements { get; set; } = [];
}

public class RecruitmentHiringCasePausePeriod
{
    public long Id { get; set; }
    public long PositionStageInstanceId { get; set; }
    public string Reason { get; set; } = "";
    public int PausedByUserId { get; set; }
    public string PausedByName { get; set; } = "";
    public DateTime PausedAtUtc { get; set; }
    public int? ResumedByUserId { get; set; }
    public string ResumedByName { get; set; } = "";
    public DateTime? ResumedAtUtc { get; set; }
    public long DurationSeconds { get; set; }
}

public class MoveRecruitmentHiringCaseRequest
{
    public string OutcomeCode { get; set; } = "ADVANCE";
    public string Reason { get; set; } = "";
}

public class RecruitmentProcessDocument
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public long? HiringCaseId { get; set; }
    public long? ApplicationId { get; set; }
    public long? InterviewId { get; set; }
    public long? PipelineStageId { get; set; }
    public string DocumentType { get; set; } = "";
    public int VersionNumber { get; set; }
    public long? TemplateId { get; set; }
    public Guid? AttachmentPublicId { get; set; }
    public bool HasFinalSignedAttachment { get; set; }
    public string Status { get; set; } = "Draft";
    public long? WorkflowInstanceId { get; set; }
    public int CreatedByUserId { get; set; }
    public int? SignedByUserId { get; set; }
    public DateTime? SignedAtUtc { get; set; }
    public DateTime CreatedAtUtc { get; set; }
    public DateTime UpdatedAtUtc { get; set; }
}

public class SaveRecruitmentProcessDocument
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public long? HiringCaseId { get; set; }
    public long? ApplicationId { get; set; }
    public long? InterviewId { get; set; }
    public long? PipelineStageId { get; set; }
    public string DocumentType { get; set; } = "";
    public long? TemplateId { get; set; }
    public Guid? AttachmentPublicId { get; set; }
    public string Status { get; set; } = "Draft";
    public long? WorkflowInstanceId { get; set; }
}

public class RecruitmentStageActionRecipient
{
    public long Id { get; set; }
    public long StageActionId { get; set; }
    public string RecipientType { get; set; } = "Candidate";
    public int? UserId { get; set; }
    public string RoleCode { get; set; } = "";
    public string EmailAddress { get; set; } = "";
    public int DisplayOrder { get; set; } = 100;
    public bool IsActive { get; set; } = true;
}

public class RecruitmentStageDefaultPanelMember
{
    public long Id { get; set; }
    public long PipelineStageId { get; set; }
    public int PanelUserId { get; set; }
    public string PanelUserName { get; set; } = "";
    public string PanelRole { get; set; } = "Panelist";
    public bool IsRequired { get; set; } = true;
    public int DisplayOrder { get; set; } = 100;
}

public class RecruitmentProfileSubmissionBatch
{
    public long Id { get; set; }
    public int ClientId { get; set; }
    public long HiringCaseId { get; set; }
    public string BatchNumber { get; set; } = "";
    public string Status { get; set; } = "Draft";
    public int CreatedByUserId { get; set; }
    public int? ApprovedByUserId { get; set; }
    public DateTime? ApprovedAtUtc { get; set; }
    public int? ForwardedByUserId { get; set; }
    public DateTime? ForwardedAtUtc { get; set; }
    public DateTime CreatedAtUtc { get; set; }
    public DateTime UpdatedAtUtc { get; set; }
    public List<RecruitmentProfileSubmissionBatchItem> Items { get; set; } = [];
    public List<RecruitmentProfileBatchNotificationDelivery> Deliveries { get; set; } = [];
}

public class RecruitmentProfileSubmissionBatchItem
{
    public long Id { get; set; }
    public long BatchId { get; set; }
    public long ApplicationId { get; set; }
    public long CandidateId { get; set; }
    public string CandidateName { get; set; } = "";
    public decimal? AtsScore { get; set; }
    public string ReadinessStatus { get; set; } = "Pending";
    public string MissingFields { get; set; } = "";
}

public class SaveRecruitmentProfileSubmissionBatch
{
    public long HiringCaseId { get; set; }
    public List<long> ApplicationIds { get; set; } = [];
}

public class RecruitmentProfileBatchNotificationDelivery
{
    public long Id { get; set; }
    public long BatchId { get; set; }
    public long StageActionId { get; set; }
    public string RecipientType { get; set; } = "";
    public string RecipientEmail { get; set; } = "";
    public long NotificationQueueId { get; set; }
    public DateTime CreatedAtUtc { get; set; }
}

File: payroll-ui/src/types/recruitmentCases.ts
export type RecruitmentWorkOrderLine = {
  id: number
  workOrderId: number
  lineNumber: number
  positionName: string
  payBandLevelCode: string
  numberOfPositions: number
  location: string
  division: string
  requisitionId?: number | null
  positionId?: number | null
  status: string
  createdAtUtc?: string
  updatedAtUtc?: string
}

export type RecruitmentWorkOrder = {
  id: number
  clientId: number
  clientName: string
  workOrderNumber: string
  receivedAtUtc: string
  receivedFrom: string
  subject: string
  remarks: string
  status: string
  overallSlaMinutes: number
  dueAtUtc?: string | null
  lineCount: number
  openCaseCount: number
  createdByUserId?: number
  createdAtUtc?: string
  updatedAtUtc?: string
  lines: RecruitmentWorkOrderLine[]
}

export type SaveRecruitmentWorkOrder = Omit<RecruitmentWorkOrder,
  'clientName' | 'dueAtUtc' | 'lineCount' | 'openCaseCount' | 'createdByUserId' | 'createdAtUtc' | 'updatedAtUtc' | 'lines'> & {
    lines: Array<Omit<RecruitmentWorkOrderLine, 'workOrderId' | 'createdAtUtc' | 'updatedAtUtc'>>
  }

export type RecruitmentHiringCasePause = {
  id: number
  positionStageInstanceId: number
  reason: string
  pausedByUserId: number
  pausedByName: string
  pausedAtUtc: string
  resumedByUserId?: number | null
  resumedByName: string
  resumedAtUtc?: string | null
  durationSeconds: number
}

export type RecruitmentHiringCaseStage = {
  id: number
  hiringCaseId: number
  pipelineStageId: number
  stageCode: string
  stageName: string
  displayOrder: number
  stakeholderCode: string
  targetOffsetMinutes?: number | null
  allowPause: boolean
  pauseBehavior: string
  requiresApproval: boolean
  approvalWorkflowId?: number | null
  isTerminal: boolean
  status: string
  enteredAtUtc?: string | null
  dueAtUtc?: string | null
  completedAtUtc?: string | null
  pausedDurationSeconds: number
  isPaused: boolean
  isSlaBreached: boolean
  pauseHistory: RecruitmentHiringCasePause[]
  processDocumentRequirements: import('./recruitmentOrchestration').RecruitmentStageProcessDocumentRequirement[]
}

export type RecruitmentHiringCase = {
  id: number
  clientId: number
  clientName: string
  workOrderId: number
  workOrderNumber: string
  workOrderLineId: number
  requisitionId?: number | null
  positionId?: number | null
  positionName: string
  payBandLevelCode: string
  division: string
  pipelineVersionId: number
  pipelineName: string
  slaAnchorAtUtc: string
  overallDueAtUtc?: string | null
  currentStageInstanceId?: number | null
  currentStageName: string
  currentStakeholderCode: string
  status: string
  advanceStatus: string
  advanceRequestId?: number | null
  advanceMessage: string
  startedAtUtc: string
  updatedAtUtc: string
  stages: RecruitmentHiringCaseStage[]
}

export type RecruitmentProcessDocument = {
  id: number
  clientId: number
  hiringCaseId?: number | null
  applicationId?: number | null
  interviewId?: number | null
  pipelineStageId?: number | null
  documentType: string
  versionNumber: number
  templateId?: number | null
  attachmentPublicId?: string | null
  hasFinalSignedAttachment: boolean
  status: string
  workflowInstanceId?: number | null
  createdByUserId: number
  signedByUserId?: number | null
  signedAtUtc?: string | null
  createdAtUtc: string
  updatedAtUtc: string
}

export type SaveRecruitmentProcessDocument = Omit<RecruitmentProcessDocument,
  'versionNumber' | 'hasFinalSignedAttachment' | 'createdByUserId' | 'signedByUserId' | 'signedAtUtc' | 'createdAtUtc' | 'updatedAtUtc'>

export type RecruitmentProfileSubmissionBatchItem = {
  id: number
  batchId: number
  applicationId: number
  candidateId: number
  candidateName: string
  applicationScoreId?: number | null
  atsScore?: number | null
  readinessStatus: string
  missingFields: string
}

export type RecruitmentProfileBatchNotificationDelivery = {
  id: number
  batchId: number
  stageActionId: number
  recipientType: string
  recipientEmail: string
  notificationQueueId: number
  createdAtUtc: string
}

export type RecruitmentProfileSubmissionBatch = {
  id: number
  clientId: number
  hiringCaseId: number
  batchNumber: string
  status: string
  createdByUserId: number
  approvedByUserId?: number | null
  approvedAtUtc?: string | null
  forwardedByUserId?: number | null
  forwardedAtUtc?: string | null
  createdAtUtc: string
  updatedAtUtc: string
  items: RecruitmentProfileSubmissionBatchItem[]
  deliveries: RecruitmentProfileBatchNotificationDelivery[]
}
